File: client_windows_native/Core/VoiceTyperController.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Threading;
using System.Threading.Tasks;
using VoiceTyper.Services;
using VoiceTyper.Support;

namespace VoiceTyper.Core;

/// <summary>
/// 中央状态机。串起热键 → 录音 → ASR → 文本插入。
/// 所有公共方法和事件回调都在 UI 线程上完成。
/// </summary>
internal sealed class VoiceTyperController : IDisposable
{
    /// <summary>
    /// 短于此时长的录音视为误触，直接丢弃：流式不发 finalize，非流式不发 HTTP 请求。
    /// 客户端侧约定，服务端不做强制校验。见 PROTOCOL.md §5.1。
    /// </summary>
    private static readonly TimeSpan MinimumRecordingDuration = TimeSpan.FromMilliseconds(300);

    public Action<AppStateInfo>? StateChanged;
    public Action<string>? PreviewUpdate;
    public Action<string>? RecognizedText;

    private readonly AppConfig _config;
    private readonly HotkeyService _hotkeyService;
    private readonly AudioCaptureService _audioService;
    private readonly TextInsertionService _textInsertion;

    private StreamingASRClient? _streamingClient;
    private List<byte[]>? _batchChunks;
    private string _accumulatedPreview = "";
    private bool _isRecording;
    private bool _isRunning;
    /// <summary>本轮录音真正开始的时间戳（Stopwatch tick），用于短录音过滤。</summary>
    private long _recordingStartedTicks;
    /// <summary>本轮录音因过短被判定丢弃；由 OnTailChunk 回调消费。</summary>
    private bool _discardCurrentSession;

    public bool IsRunning => _isRunning;

    public VoiceTyperController(AppConfig config)
    {
        _config = config;
        _hotkeyService = new HotkeyService();
        _audioService = new AudioCaptureService();
        _textInsertion = new TextInsertionService();
    }

    public void Start()
    {
        if (_isRunning) return;

        _hotkeyService.OnPress = BeginRecording;
        _hotkeyService.OnRelease = FinishRecording;
        _hotkeyService.Start(_config.Hotkey);

        _isRunning = true;
        StateChanged?.Invoke(AppStateInfo.Idle);
        AppLog.Info("controller", "Controller started");
    }

    public void Stop()
    {
        if (!_isRunning) return;
        _isRunning = false;

        _hotkeyService.Stop();
        _audioService.StopWithoutResult();
        TeardownStreamingClient();
        _batchChunks = null;
        _isRecording = false;
        _discardCurrentSession = false;
        AppLog.Info("controller", "Controller stopped");
    }

    public void Dispose()
    {
        Stop();
        _hotkeyService.Dispose();
        _audioService.Dispose();
    }

    public Task<bool> HealthCheckAsync(CancellationToken ct = default) =>
        ASRClient.HealthCheckAsync(_config.Server, ct);

    // ─── 录音生命周期 ─────────────────────────────────────────────

    private void BeginRecording()
    {
        if (!_isRunning || _isRecording) return;
        // 上一轮的丢弃标记若因异常路径未被 OnTailChunk 消费，这里兜底清掉。
        _discardCurrentSession = false;
        if (_config.Server.Streaming) BeginStreamingRecording();
        else BeginBatchRecording();
    }

    private void FinishRecording()
    {
        if (!_isRecording) return;
        _isRecording = false;

        // 短录音（误触）在这里就判定，交给 OnTailChunk 执行丢弃：
        // 走同一条收尾路径，避免录音流没被正常停掉。
        var elapsed = Stopwatch.GetElapsedTime(_recordingStartedTicks);
        _discardCurrentSession = elapsed < MinimumRecordingDuration;
        if (_discardCurrentSession)
        {
            AppLog.Info("controller", $"录音过短（{elapsed.TotalMilliseconds:F0}ms），已丢弃");
        }

        // 录音停止 → 触发 OnTailChunk → 流式：sendAudio(tail) + finalize；非流式：拼接整段后 POST
        _audioService.Stop();
    }

    /// <summary>
    /// 丢弃本轮过短录音的共同收尾：清预览、断连接、回到就绪。
    /// <paramref name="client"/> 为流式会话的连接，非流式路径传 null。
    /// </summary>
    private void DiscardShortSession(StreamingASRClient? client)
    {
        _discardCurrentSession = false;

        // 会话可能已被新一轮录音取代：只关掉自己的连接，不触碰当前会话的状态。
        if (client is not null && !ReferenceEquals(_streamingClient, client))
        {
            client.Close();
            return;
        }

        _accumulatedPreview = "";
        PreviewUpdate?.Invoke("");
        TeardownStreamingClient();
        if (!_isRecording) StateChanged?.Invoke(AppStateInfo.Idle);
    }

    // ─── 流式路径 ────────────────────────────────────────────────

    private void BeginStreamingRecording()
    {
        var client = new StreamingASRClient();

        // partial 是全量预览文本，直接替换（协议 v2，见 PROTOCOL.md §4.3）。
        // 服务端整段重跑会修正先前的文字，所以不能再做拼接。
        client.OnPartial = text =>
        {
            _accumulatedPreview = text;
            PreviewUpdate?.Invoke(_accumulatedPreview);
        };

        // 通过对象引用比较判断是否为当前会话；不是的话静默插入并关闭旧连接，
        // 不触碰当前会话的状态。
        client.OnFinal = text =>
        {
            if (ReferenceEquals(_streamingClient, client))
            {
                _accumulatedPreview = "";
                PreviewUpdate?.Invoke("");
                TeardownStreamingClient();
                HandleFinalText(text);
            }
            else
            {
                client.Close();
                var trimmed = (text ?? "").Trim();
                if (!_isRunning || string.IsNullOrEmpty(trimmed)) return;
                _textInsertion.Insert(trimmed);
                RecognizedText?.Invoke(trimmed);
            }
        };

        client.OnError = message =>
        {
            AppLog.Error("controller", $"ASR error: {message}");
            if (ReferenceEquals(_streamingClient, client))
            {
                TeardownStreamingClient();
                _accumulatedPreview = "";
                PreviewUpdate?.Invoke("");
                StateChanged?.Invoke(AppStateInfo.ErrorWith(message));
                _isRecording = false;
            }
            else
            {
                client.Close();
            }
        };

        // 先连接 WebSocket（异步），连接成功后再启动音频。
        // 连接是同步等待的（块在调用线程），但因为我们在 UI 线程，
        // 用 fire-and-forget Task.Run 避免卡 UI；启动录音放到连接完成的延续里。
        _streamingClient = client;
        _accumulatedPreview = "";

        _ = ConnectAndStartCaptureAsync(client);
    }

    private async Task ConnectAndStartCaptureAsync(StreamingASRClient client)
    {
        try
        {
            await client.ConnectAsync(_config.Server, _config.Server.LlmRecorrect).ConfigureAwait(true);
        }
        catch (Exception ex)
        {
            AppLog.Error("controller", "WebSocket 连接失败", ex);
            UiDispatcher.Post(() =>
            {
                if (ReferenceEquals(_streamingClient, client))
                {
                    TeardownStreamingClient();
                    StateChanged?.Invoke(AppStateInfo.ErrorWith("无法连接到识别服务"));
                }
                else
                {
                    client.Close();
                }
            });
            return;
        }

        UiDispatcher.Post(() => StartAudioCaptureForStreaming(client));
    }

    private void StartAudioCaptureForStreaming(StreamingASRClient client)
    {
        // 连接到位时，会话可能已经被新的录音覆盖（用户连按两次）。
        if (!ReferenceEquals(_streamingClient, client))
        {
            client.Close();
            return;
        }

        // 工作线程回调 → UI 线程后再 sendAudio（StreamingASRClient 自身是线程安全的，但保持模型一致）
        _audioService.OnChunk = data =>
        {
            UiDispatcher.Post(() => client.SendAudio(data));
        };
        _audioService.OnTailChunk = data =>
        {
            UiDispatcher.Post(() =>
            {
                // 误触录音：不发 finalize，直接断开连接。服务端会随连接关闭丢掉会话。
                if (_discardCurrentSession)
                {
                    DiscardShortSession(client);
                    return;
                }

                if (data.Length > 0) client.SendAudio(data);
                client.FinalizeStream();
                StateChanged?.Invoke(AppStateInfo.Recognizing);
            });
        };

        try
        {
            _audioService.Start();
        }
        catch (AudioStartException ex)
        {
            AppLog.Error("controller", "启动录音失败", ex);
            client.Close();
            _streamingClient = null;
            var msg = ex.IsAccessDenied ? "麦克风权限被拒绝，请在 Windows 设置中允许应用访问麦克风" : "开始录音失败";
            StateChanged?.Invoke(AppStateInfo.ErrorWith(msg));
            return;
        }

        // 计时从音频真正开始采集起算，不含 WebSocket 连接耗时。
        _recordingStartedTicks = Stopwatch.GetTimestamp();
        _isRecording = true;
        StateChanged?.Invoke(AppStateInfo.Recording);
    }

    // ─── 非流式路径 ──────────────────────────────────────────────

    private void BeginBatchRecording()
    {
        _batchChunks = new List<byte[]>();

        _audioService.OnChunk = data =>
        {
            UiDispatcher.Post(() => _batchChunks?.Add(data));
        };

        _audioService.OnTailChunk = data =>
        {
            UiDispatcher.Post(() =>
            {
                // 误触录音：不发出 HTTP 请求，直接丢掉已累积的音频。
                if (_discardCurrentSession)
                {
                    DiscardShortSession(null);
                    return;
                }

                if (data.Length > 0) _batchChunks?.Add(data);
                _ = PerformBatchRecognitionAsync();
            });
        };

        try
        {
            _audioService.Start();
        }
        catch (AudioStartException ex)
        {
            AppLog.Error("controller", "启动录音失败", ex);
            var msg = ex.IsAccessDenied ? "麦克风权限被拒绝，请在 Windows 设置中允许应用访问麦克风" : "开始录音失败";
            StateChanged?.Invoke(AppStateInfo.ErrorWith(msg));
            return;
        }

        _recordingStartedTicks = Stopwatch.GetTimestamp();
        _isRecording = true;
        StateChanged?.Invoke(AppStateInfo.Recording);
    }

    private async Task PerformBatchRecognitionAsync()
    {
        StateChanged?.Invoke(AppStateInfo.Recognizing);

        var chunks = _batchChunks ?? new List<byte[]>();
        _batchChunks = null;

        if (chunks.Count == 0)
        {
            StateChanged?.Invoke(AppStateInfo.Idle);
            return;
        }

        // 拼接为单段 float32 PCM
        int total = 0;
        foreach (var c in chunks) total += c.Length;
        var combined = new byte[total];
        int offset = 0;
        foreach (var c in chunks)
        {
            Buffer.BlockCopy(c, 0, combined, offset, c.Length);
            offset += c.Length;
        }


        try
        {
            using var client = new ASRClient(_config.Server);
            var text = await client.RecognizeAsync(combined, _config.Server.LlmRecorrect).ConfigureAwait(true);
            HandleFinalText(text);
        }
        catch (Exception ex)
        {
            AppLog.Error("controller", "批量识别失败", ex);
            StateChanged?.Invoke(AppStateInfo.ErrorWith($"识别失败：{ex.Message}"));
        }
    }

    // ─── 公共处理 ────────────────────────────────────────────────

    private void HandleFinalText(string text)
    {
        var trimmed = (text ?? "").Trim();

        if (string.IsNullOrEmpty(trimmed))
        {
            if (!_isRecording) StateChanged?.Invoke(AppStateInfo.Idle);
            return;
        }

        if (!_isRunning) return;

        StateChanged?.Invoke(AppStateInfo.Inserting);
        var inserted = _textInsertion.Insert(trimmed);
        if (inserted)
        {
            RecognizedText?.Invoke(trimmed);
            // 若此时已有新一轮录音正在进行，不要把状态拉回 Idle
            if (!_isRecording) StateChanged?.Invoke(AppStateInfo.Idle);
        }
        else
        {
            StateChanged?.Invoke(AppStateInfo.ErrorWith("文本插入失败"));
        }
    }

    private void TeardownStreamingClient()
    {
        _streamingClient?.Close();
        _streamingClient = null;
        _audioService.OnChunk = null;
        _audioService.OnTailChunk = null;
        _batchChunks = null;
    }
}

File: client_macos_swift/Sources/VoiceTyperApp/Core/VoiceTyperController.swift
import AppKit
import Foundation

@MainActor
final class VoiceTyperController {
    private let config: AppConfig
    private let hotwords: [String]
    private let hotkeyService: HotkeyService
    private let audioCaptureService: AudioCaptureService
    private let textInsertionService: TextInsertionService

    private var asrClient: StreamingASRClient?
    private var isRecording = false
    private var isRunning = false
    private var accumulatedPreview = ""
    private var batchAudioChunks: [Data] = []
    private var recordingStartedAt: Date?

    /// 录音时长低于此阈值的会话直接丢弃，避免误触上传无意义音频。
    /// 见仓库根目录 PROTOCOL.md "短录音过滤"。
    private static let minimumRecordingDuration: TimeInterval = 0.3

    var onStateChange: ((AppState) -> Void)?
    var onRecognizedText: ((String) -> Void)?
    var onPreviewUpdate: ((String) -> Void)?
    /// 非致命提示（流式 partial 失败等），UI 可短暂闪烁状态但不打断录音。
    var onPreviewWarning: ((String) -> Void)?
    /// 用户主动取消（录音中按 Esc）。与 .idle 区分，便于 UI 给出"已取消"提示。
    var onCancelled: (() -> Void)?
    var isStarted: Bool { isRunning }

    init(
        config: AppConfig,
        hotwords: [String],
        hotkeyService: HotkeyService = HotkeyService(),
        audioCaptureService: AudioCaptureService = AudioCaptureService(),
        textInsertionService: TextInsertionService = TextInsertionService()
    ) {
        self.config = config
        self.hotwords = hotwords
        self.hotkeyService = hotkeyService
        self.audioCaptureService = audioCaptureService
        self.textInsertionService = textInsertionService
    }

    func start() throws {
        guard !isRunning else { return }

        hotkeyService.onPress = { [weak self] in
            Task { @MainActor [weak self] in self?.beginRecording() }
        }
        hotkeyService.onRelease = { [weak self] in
            Task { @MainActor [weak self] in self?.finishRecording() }
        }
        hotkeyService.onCancel = { [weak self] in
            Task { @MainActor [weak self] in self?.cancelByUser() }
        }
        try hotkeyService.start(with: config.hotkey)
        isRunning = true
        // 预热音频引擎（不开麦、不亮指示灯），缩短首次按键的冷启动。
        audioCaptureService.prewarm()
        onStateChange?(.idle)
    }

    func stop() {
        hotkeyService.stop()
        audioCaptureService.stopWithoutResult()
        teardownASRClient()
        isRunning = false
        isRecording = false
    }

    func healthCheck() async -> Bool {
        let result = await ServerHealthProbe.check(server: config.server)
        if let version = result.version {
            AppLog.network.info("服务端版本: \(version, privacy: .public)")
        }
        return result.ready
    }

    // MARK: - 录音流程

    private func beginRecording() {
        guard isRunning, !isRecording else { return }

        recordingStartedAt = Date()
        if config.server.streaming {
            beginStreamingRecording()
        } else {
            beginBatchRecording()
        }
    }

    private func finishRecording() {
        guard isRecording else { return }
        isRecording = false

        // 短录音过滤：低于阈值的录音视为误触，立即取消。
        if let startedAt = recordingStartedAt,
           Date().timeIntervalSince(startedAt) < Self.minimumRecordingDuration {
            AppLog.audio.info("录音时长低于阈值，已丢弃")
            cancelCurrentRecording()
            return
        }

        // 流式：stop() 触发 onTailChunk → sendAudio(tail) + finalize()
        // 非流式：stop() 触发 onTailChunk → 累积尾音 → HTTP POST
        audioCaptureService.stop()
    }

    /// 停止采集、丢弃尾音、关闭 WS、清空预览并重新预热引擎。
    /// 不做状态转换，由调用方决定回到 idle 还是发"已取消"提示。
    private func resetRecording() {
        audioCaptureService.stopWithoutResult()
        teardownASRClient()
        accumulatedPreview = ""
        onPreviewUpdate?("")
        recordingStartedAt = nil
        isRecording = false
        audioCaptureService.prewarm()
    }

    /// 取消当前录音并静默回到 idle（短录音过滤等内部触发）。
    private func cancelCurrentRecording() {
        resetRecording()
        if isRunning {
            onStateChange?(.idle)
        }
    }

    /// 用户在录音过程中按 Esc 主动取消，通过 `onCancelled` 让 UI 给出"已取消"提示。
    private func cancelByUser() {
        guard isRecording else { return }
        resetRecording()
        if isRunning {
            AppLog.audio.info("用户取消录音")
            onCancelled?()
        }
    }

    // MARK: - 流式路径

    private func beginStreamingRecording() {
        let client = StreamingASRClient()

        client.onPartial = { [weak self] fragment in
            guard let self else { return }
            self.accumulatedPreview += fragment
            self.onPreviewUpdate?(self.accumulatedPreview)
        }

        // [weak client] 而非 self.asrClient：onFinal 触发时 asrClient 可能已指向更新的会话，
        // 通过身份比较确认是否是当前会话，避免误关后续录音的 WS 连接。
        client.onFinal = { [weak self, weak client] text in
            guard let self else { return }
            if let c = client, self.asrClient === c {
                // 当前（最新）会话完成：清理预览、完整拆解、状态转换
                self.accumulatedPreview = ""
                self.onPreviewUpdate?("")
                self.teardownASRClient()
                self.handleFinalText(text)
            } else {
                // 旧会话在后台完成：静默关闭 + 直接插入，不影响当前会话状态
                client?.close()
                let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
                guard !trimmed.isEmpty, self.isRunning else { return }
                _ = self.textInsertionService.insert(text: trimmed)
                self.onRecognizedText?(trimmed)
            }
        }

        client.onWarning = { [weak self, weak client] message in
            guard let self else { return }
            // 仅当前会话的 warning 才转发到 HUD；旧会话的提示直接丢弃。
            if let c = client, self.asrClient === c {
                self.onPreviewWarning?(message)
            }
        }

        client.onError = { [weak self, weak client] message in
            guard let self else { return }
            AppLog.network.error("ASR 错误: \(message, privacy: .public)")
            if let c = client, self.asrClient === c {
                // 停止采集，避免 WS 中途出错后 AVAudioEngine 继续运行、麦克风指示灯常亮，
                // 并防止残留音频混入下一次录音。
                self.audioCaptureService.stopWithoutResult()
                self.teardownASRClient()
                self.accumulatedPreview = ""
                self.onPreviewUpdate?("")
                self.isRecording = false
                self.audioCaptureService.prewarm()
                self.onStateChange?(.error(message))
            } else {
                // 旧会话出错，不干扰当前会话
                client?.close()
            }
        }

        do {
            try client.connect(
                server: config.server,
                hotwords: hotwords,
                llmRecorrect: config.server.llmRecorrect
            )
        } catch {
            AppLog.network.error("WS 连接失败: \(error.localizedDescription, privacy: .public)")
            onStateChange?(.error("无法连接到识别服务"))
            return
        }

        // onChunk 在 AVAudioEngine 音频线程回调，需跳回 MainActor 再访问 @MainActor 的 client
        audioCaptureService.onChunk = { [weak client] data in
            Task { @MainActor [weak client] in
                client?.sendAudio(data)
            }
        }

        // onTailChunk 同理；sendAudio + finalize 需保证在同一 MainActor 执行块内顺序完成
        audioCaptureService.onTailChunk = { [weak self, weak client] data in
            Task { @MainActor [weak self, weak client] in
                guard let self else { return }
                if !data.isEmpty {
                    client?.sendAudio(data)
                }
                // 传入超时：服务端卡住 / 网络半开时不至于让 HUD 永久停在"识别中"。
                client?.finalize(timeout: self.config.server.timeout)
                self.onStateChange?(.recognizing)
            }
        }

        do {
            try audioCaptureService.start()
        } catch {
            AppLog.audio.error("开始录音失败: \(error.localizedDescription, privacy: .public)")
            client.close()
            onStateChange?(.error("开始录音失败"))
            return
        }

        asrClient = client
        isRecording = true
        accumulatedPreview = ""
        onStateChange?(.recording)
    }

    // MARK: - 非流式路径

    private func beginBatchRecording() {
        batchAudioChunks = []

        audioCaptureService.onChunk = { [weak self] data in
            self?.batchAudioChunks.append(data)
        }

        audioCaptureService.onTailChunk = { [weak self] data in
            Task { @MainActor [weak self] in
                guard let self else { return }
                if !data.isEmpty {
                    self.batchAudioChunks.append(data)
                }
                await self.performBatchRecognition()
            }
        }

        do {
            try audioCaptureService.start()
        } catch {
            AppLog.audio.error("开始录音失败: \(error.localizedDescription, privacy: .public)")
            onStateChange?(.error("开始录音失败"))
            return
        }

        isRecording = true
        onStateChange?(.recording)
    }

    private func performBatchRecognition() async {
        let combinedData = batchAudioChunks.reduce(Data(), +)
        batchAudioChunks = []

        // 兜底：若实际音频不足阈值（采样数 = bytes / 4），同样丢弃。
        let sampleCount = combinedData.count / MemoryLayout<Float>.size
        let minimumSamples = Int(Self.minimumRecordingDuration * AppConstants.targetSampleRate)
        guard !combinedData.isEmpty, sampleCount >= minimumSamples else {
            teardownASRClient()
            audioCaptureService.prewarm()
            onStateChange?(.idle)
            return
        }

        onStateChange?(.recognizing)

        let client = ASRClient(server: config.server)
        let hotwordsString = hotwords.joined(separator: " ")

        do {
            let text = try await client.recognize(
                audioData: combinedData,
                hotwords: hotwordsString,
                llmRecorrect: config.server.llmRecorrect
            )
            teardownASRClient()
            handleFinalText(text)
        } catch {
            teardownASRClient()
            audioCaptureService.prewarm()
            AppLog.network.error("批量识别失败: \(error.localizedDescription, privacy: .public)")
            onStateChange?(.error("识别失败：\(error.localizedDescription)"))
        }
    }

    // MARK: - 公共处理

    /// 插入最终文本并更新状态。
    /// 调用方负责在调用前完成会话拆解（teardownASRClient / client.close）。
    private func handleFinalText(_ text: String) {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)

        guard !trimmed.isEmpty else {
            if !isRecording {
                audioCaptureService.prewarm()
                onStateChange?(.idle)
            }
            return
        }

        guard isRunning else { return }

        onStateChange?(.inserting)
        let inserted = textInsertionService.insert(text: trimmed)
        if inserted {
            onRecognizedText?(trimmed)
            // 若此时有新录音正在进行（并发会话），不覆盖 .recording 状态
            if !isRecording {
                audioCaptureService.prewarm()
                onStateChange?(.idle)
            }
        } else {
            // 插入失败兜底：把结果写入剪贴板，避免长听写内容彻底丢失。
            textInsertionService.copyToClipboard(text: trimmed)
            AppLog.app.error("文本插入失败，已复制到剪贴板")
            onStateChange?(.error("插入失败，已复制到剪贴板，可手动粘贴"))
        }
    }

    private func teardownASRClient() {
        asrClient?.close()
        asrClient = nil
        batchAudioChunks = []
        audioCaptureService.onChunk = nil
        audioCaptureService.onTailChunk = nil
    }
}
